File: ex04/task1/task1.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>


struct Person {
	std::string first_name;
	std::string last_name;
	int age;
};

bool operator ==(const Person& lhs, const Person& rhs) {
	return lhs.first_name == rhs.first_name
	       && lhs.last_name == rhs.last_name && lhs.age == rhs.age;
}

class Room {
public:
	Room(int id, size_t limit) : id{id}, limit{limit} {}
	// Returns true iff the person successfully entered.
	bool enter(Person person) {
		if(people.size() < limit) {
			people.push_back(person);
			std::cout << "Person " << person.first_name << " "
			<< person.last_name << " entered Room with ID: " << id << std::endl;
			return true;
		}
		std::cout << "Person " << person.first_name << " "
		<< person.last_name << " was not able to enter Room with ID: " << id
		<< " because it's person limit is reached" << std::endl;
		return false;
	}
	void exit(Person person) {
		std::cout << "Person " << person.first_name << " "
		<< person.last_name << " left Room with ID: " << id << std::endl;
		people.erase(std::remove(people.begin(), people.end(), person), people.end());
	}

	size_t getNumberOfPeople() const {
		return people.size();
	}

	bool isInRoom(Person person) const {
		if(std::find(people.begin(), people.end(), person) != people.end()){
			std::cout << "Person " << person.first_name << " "
			<< person.last_name << " is located in Room with ID: " << id << std::endl;
			return true;
		}
		std::cout << "Person " << person.first_name << " "
		<< person.last_name << " is not located in Room with ID: " << id << std::endl;
		return false;
	}

	std::vector<Person> getPeople() const {
		return people;
	}

private:
	const int id;
	const size_t limit;
	std::vector<Person> people;
};


int main(int argc, char * argv[]){

	Person p1;
	p1.first_name = "Max";
	p1.last_name = "Mustermann";
	p1.age = 21;

	Room r1{1, 5};
	r1.enter(p1);
	r1.isInRoom(p1);
	r1.exit(p1);
	r1.isInRoom(p1);

	Room r2{2, 1};
	Person p2;
	p2.first_name = "Muster";
	p2.last_name = "Maxmann";
	p2.age = 22;

	r2.enter(p1);
	r2.enter(p2);
}

File: ex04/task1/Task1.java
import java.util.ArrayList;
import java.util.List;

public class Task1 {

  public static void main(String[] args) {
      Person p1 = new Person("Klaus", "Test", 46);
      Room r = new Room(1, 10);

      r.iterateOverPeople();
      r.enter(p1);
      r.iterateOverPeople();
      r.exit(p1);
      r.iterateOverPeople();
  }

}

class Person {

  private String firstName;
  private String lastName;
  private int age;

  public Person() {}

  public Person(String firstName, String lastName, int age){
    this.firstName = firstName;
    this.lastName = lastName;
    this.age = age;
  }

  public String getInfo() {
    return this.firstName+ " " + this.lastName+ " ("+ this.age +")";
  }
}

class Room {

  private int id;
  // Java has no unsigned integers
  private int limit;
  private List<Person> people = new ArrayList<>();

  public Room(int id, int limit) {
    if(limit < 0) {
      // should throw exception
      this.limit = 0;
    } else {
      this.id = id;
      this.limit = limit;
    }
  }

  public int countPeople(){
    return this.people.size();
  }

  public boolean isPersonInRoom(Person p){
    return this.people.contains(p);
  }

  public void iterateOverPeople(){
    if(this.people.isEmpty()){
      System.out.println("Room empty");
      return;
    }
    System.out.println("People in room:");
    for(Person p : this.people){
  		System.out.print(p.getInfo() + ", ");
  	}
    System.out.println("");
  }

  public boolean enter(Person p){
    if(this.people.size() < this.limit)
    {
      this.people.add(p);
      System.out.println(p.getInfo() + " has entered the room");
      return true;
    }
    System.out.println("Room is full! " + p.getInfo() + " cannot enter");
    return false;
  }

  public void exit(Person p){
    this.people.remove(p);
  }

}
